Added the ability to run and combine multiple animations simultaneously

mohammadbaghaei opened the pull request:
Added the ability to run and combine multiple animations simultaneously with `appendAnimation` and `detachAnimation` methods, updated documentation, upgrade to three.js r171 and update all dependencies to latest version

<!-- Instructions: https://github.com/google/model-viewer/blob/master/CONTRIBUTING.md#code-reviews -->

elalish commented on packages/model-viewer/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@google/model-viewer",
-  "version": "4.0.0",
+  "version": "5.0.0",
```
Thanks for all the dependency updates! I don't believe any of this will be breaking changes though, correct? So no need for a major version bump. Also, we do version bumps of our packages when we prepare a release, not during regular PRs, so you can revert this.

elalish commented on packages/model-viewer/src/features/animation.ts:
```diff
@@ -13,29 +13,57 @@
 
-import {property} from 'lit/decorators.js';
-import {LoopOnce, LoopPingPong, LoopRepeat} from 'three';
+import { property } from 'lit/decorators.js';
```
It looks like your formatter isn't picking up our `.clang-format` file? Can you get it to keep consistent formatting with our files?

elalish commented on packages/modelviewer.dev/data/docs.json:
```diff
@@ -1023,6 +1028,16 @@
+      {
+        "name": "appendAnimation(animationName, options: {repetitions, pingpong, weight, timeScale, fade, warp})",
+        "htmlName": "appendAnimation",
+        "description": "Causes an animation to run simultaneously with and blend into the current animation. If you want to get the appended animations, you can use <span class='attribute'>appendedAnimations</span> property. You can specify the number of repetitions of the animation by setting the number of <i>repetitions</i> to any value greater than 0 (defaults to Infinity). Also if you set <i> ...
```
Thanks for the detailed docs! Since this is a fairly complex feature, can you also add a few automated tests and an example? We use our examples as our manual testing plan, and I think it will help users quite a bit to see this API in action.

elalish commented on packages/modelviewer.dev/data/docs.json:
```diff
@@ -1045,6 +1060,16 @@
+      },
+      {
+        "name": "detachanimation",
+        "htmlName": "detachanimation",
```
These should be kabob-case like `model-visibility`.

elalish commented on packages/render-fidelity-tools/scripts/compare-fidelity-to-last-recorded.sh:
Any idea what these diffs are?

mohammadbaghaei replied:
I think it's because of the metadata change due to using the chmod command.

elalish reviewed (changes requested):
Wow, this is quite a feature, thank you! To make this easier to review and merge, would you be so kind as to split the dependency updates into a separate PR from the animation work? We should be able to quickly merge the updates, but we'll probably need some back-and-forth on the API change. Honestly, it's looking really good though!

mohammadbaghaei commented:
Hey there! Here's my rework of that PR message:

👋 Thank you for the helpful feedback! I've made the following changes in this PR:

- Fixed all the mentioned issues
- Enhanced modelviewer.dev styles for better presentation, especially in the examples section
- Updated dependencies in a separate commit
- Improved appendAnimation and detachAnimation methods to automatically handle invalid parameters 
- Added automatic parameter type conversion - numbers and booleans can now be passed as strings (case-insensitive)
- For better AR support across different devices (ref: issue #4960), moved hit-test to optionalFeatures

Let me know if any further adjustments are needed! 🙂

elalish commented:
Thanks! As this is my last day (see #4975), I'm going to pass this on to one our other maintainers, @diegoteran, who I already talked to about this PR. Looks like a promising feature!

diegoteran commented:
@mohammadbaghaei Hey, this is great! Give me a few days and then I'll review it (scheduling and deciding things with the team now that Emmett is not here) and I'll merge it to have it ready for the next MV release.

Thank you so much!
